hasura: 1.2.1 -> 1.3.1-beta.1

ingenieroariel opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
This upgrades Hasura to use GHC 8.10 in anticipation of the 20.09 release. I was pleasantly surprised when I found it available on 20.03 and sad when it went missing from nixos-unstable.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ x ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

ingenieroariel committed:
hasura: 1.2.1 -> 1.3.1-beta.1

ingenieroariel commented:
Ping @offlinehacker @peti

ingenieroariel commented:
It was very hard to track this with GHC86 so I just looked for the first version that supported GHC810 and was able to get it to work with some small mods.

There is a problem with infinite recursion when trying to override some packages like time, so I agree with @peti 's analysis that this recipe is brittle, but on the other hand, since it is an application and not a library I'd say it is okay to pin some of its dependencies.

peti commented:
Hazura should not duplicate all kinds of builds from 'haskellPackages'.

peti reviewed (changes requested):
Please don't duplicate code that we already have in `haskellPackages`. We have a very sophisticated infrastructure for packaging Haskell tools and it's a little frustrating that Hazura re-invents the wheel here instead of using the existing solutions.

offlinehacker commented:
@peti can you then suggest what would be the way to package hasura (it's not on hackage and uses very specific version constraints for some packages)?

ingenieroariel commented:
I am happy to put the work to package it differently, what I did at first was just understand what @offlinehacker had done and update it to latest GHC and Hasura. The dance to get the packages just right is very finicky so there would have to be some exceptions where we pull different versions from Hackage to not need to alter Hasura itself.

Here are a few specific examples:

1. optics-core, the version we get by default is not useful for Hasura, but currently nixpkgs also has `optics-core_0_3_0_1` and we were able to do something like:

`optics-core = super.optics-core_0_3_0_1;`

2. cron, the version we get by default is not useful for Hasura, if we update it nixpkgs wide then other things could break, so we created a local file `cron.nix` with
```
mkDerivation {
  pname = "cron";
  version = "0.7.0";
  sha256 = "sha256-pc1Ixv0RlDCqj70FllSAbw85FkLfA1FtmzAh369YEjk=";
  ...
}
```
and then used it like:
`cron = self.callPackage ./cron.nix { };`

3. The package was specifying bounds but it really wasn't necessary since it compiled with the newer ones:
`superbuffer = dontCheck (doJailbreak (unmarkBroken super.superbuffer));`

Notice that this uses the advanced machinery already in Nixpkgs for Haskell (all credit goes to the original author) so it is not necessary reinventing the wheel, but I do agree that it would be better to do away with the `cron.nix` pattern if we can find a better way to override just version numbers, I just need some direction.

ingenieroariel commented:
@peti  Here is some example code from GHC 8.10 currently on master that seems to use two of the patterns used in hasura.nix:

```
  # Deviate from Stackage LTS-15.x to fix the build.
  haddock-library = self.haddock-library_1_9_0;

  # Jailbreak to fix the build.
  base-noprelude = doJailbreak super.base-noprelude;
  pandoc = doJailbreak super.pandoc;
  system-fileio = doJailbreak super.system-fileio;
  unliftio-core = doJailbreak super.unliftio-core;

  # Use the latest version to fix the build.
  dhall = self.dhall_1_34_0;
  lens = self.lens_4_19_2;
  optics-core = self.optics-core_0_3_0_1;
  repline = self.repline_0_4_0_0;
  singletons = self.singletons_2_7;
  th-desugar = self.th-desugar_1_11;
```

https://github.com/NixOS/nixpkgs/blob/99ec1661420a1c2cba1e012380dd6344bb640799/pkgs/development/haskell-modules/configuration-ghc-8.10.x.nix#L60

ingenieroariel commented:
Looked at other packages here and found this, instead of making the versions `Reflex` wanted available in Nixpkgs, this pulled patches made to the upstream to support the versions in Nixpkgs. This may require help from someone with better Haskell chops than mine, but it is not out of the question:

```
  reflex = appendPatches super.reflex [
    # Upstream PR: https://github.com/reflex-frp/reflex/pull/434
    # Bump version bounds
    (pkgs.fetchpatch {
      url = "https://github.com/reflex-frp/reflex/commit/e6104bdfd7f664f524b6765275490722e376df4d.patch";
      sha256 ="1awp5p4640cnhfd50dplsvp0kzy6h8r0hpbw1s40blni74r3dhzr";
    })
    # Upstream PR: https://github.com/reflex-frp/reflex/pull/436
    # Fix build with newest dependent-map version
    (pkgs.fetchpatch {
      url = "https://github.com/reflex-frp/reflex/commit/dc3bf44d822d70594e3c474fe3869261776c3554.patch";
      sha256 ="0rbjfj9b8p6zkvd5j4pak5kpgard6cyfvzk750s4xwpc1v84iiqd";
    })
    # Upstream PR: https://github.com/reflex-frp/reflex/pull/437
    # Fix tests with newer dep versions
    (pkgs.fetchpatch {
      url = "https://github.com/reflex-frp/reflex/commit/87c74a1b9d9098eae8a56148c59ed4963a5232c2.patch";
      sha256 ="0qhjjgd6n4fms1hpbblny78c95bfh74izhx9dvrdlnhz6q7xlm9q";
    })
  ];
```

https://github.com/NixOS/nixpkgs/blame/99ec1661420a1c2cba1e012380dd6344bb640799/pkgs/development/haskell-modules/configuration-common.nix#L1252

ingenieroariel commented:
I refactored the code to have graphql-engine and other haskell packages correctly as libraries and made the code that compiles the application consist of only jailbreaks and small mods, it is in essence the same code that was running before but makes the packages available for others.

@peti Does this satisfy your concerns?

offlinehacker commented:
@ingenieroariel how did you generate `hackage-packages.nix`, since I dont see any changes to `configuration-hackage2nix.yaml` ?

ingenieroariel commented:
Ha, manually of course, and after reading your question realizing that was the wrong way ...

... back to reading mode :)

offlinehacker commented:
The issue I see is that hackage2nix can only generate packages from hackage as far as I am aware.

@peti is there any way to use git repo for a package not published on hackage when using hackage2nix?

ingenieroariel commented:
Ok, so now the process seems to be:

#1. Create a new PR with the following change 
#2 Wait a day for it to be applied
#3 Update this PR and take the hasura packages out again and back to the folder where they were?:
Is there a "manual haskell packages" file where I could put them instead?

```
diff --git a/pkgs/development/haskell-modules/configuration-hackage2nix.yaml b/pkgs/development/haskell-modules/configuration-hackage2nix.yaml
index 586415a1db8..715f14c5b95 100644
--- a/pkgs/development/haskell-modules/configuration-hackage2nix.yaml
+++ b/pkgs/development/haskell-modules/configuration-hackage2nix.yaml
@@ -2611,6 +2611,9 @@ extra-packages:
   - yesod-persistent < 1.5              # pre-lts-11.x versions neeed by git-annex 6.20180227
   - yesod-static ^>= 1.5                # pre-lts-11.x versions neeed by git-annex 6.20180227
   - yesod-test ^>= 1.5                  # pre-lts-11.x versions neeed by git-annex 6.20180227
+  - immortal == 0.2.2.1                 # required by Hasura
+  - dependent-map == 0.2.4.0            # required by Hasura
+  - dependent-sum == 0.4.0              # required by Hasura

 package-maintainers:
   peti:
```

offlinehacker commented:
In `non-hackage-packages.nix` move to separate files, also rebase. But thanks, this is getting better 🙂

ingenieroariel commented:
I can move it to separate files, but in which folder? The same one?

offlinehacker commented:
I would just put them in hasura folder for now.

maralorn commented on pkgs/development/haskell-modules/configuration-hackage2nix.yaml:
```diff
@@ -2611,6 +2611,9 @@ extra-packages:
   - yesod-persistent < 1.5              # pre-lts-11.x versions neeed by git-annex 6.20180227
   - yesod-static ^>= 1.5                # pre-lts-11.x versions neeed by git-annex 6.20180227
   - yesod-test ^>= 1.5                  # pre-lts-11.x versions neeed by git-annex 6.20180227
+  - immortal == 0.2.2.1                 # required by Hasura
```
Can you annotate a hasura version here, so that we can have a ball park when to remove this? (also adding a date might make sense, but the other comments don‘t do either, so that's not a must.)

maralorn commented on pkgs/servers/hasura/default.nix:
```diff
@@ -25,30 +24,8 @@ let
-      ghc-heap-view = disableLibraryProfiling (self.callPackage ./ghc-heap-view.nix { });
-
-      # unmark broewn packages and do required modifications
+      # unmark broken packages and do required modifications
```
These overrides are much better suited for the `pkgs/development/haskell-modules/configuration-common.nix` file.

Also the use of `unmarkBroken` is not the preferred approach here. If those packages actually compile, they should be marked unbroken by removing them from the broken list in `configuration-hackage2nix.yaml` so that everyone using nixpkgs can benefit from those packages.

maralorn commented on pkgs/development/haskell-modules/hackage-packages.nix:
```diff
@@ -143332,7 +143373,7 @@ self: {
 
   "insert-ordered-containers" = callPackage
     ({ mkDerivation, aeson, base, base-compat, hashable, lens
-     , optics-core, optics-extra, QuickCheck, semigroupoids, semigroups
```
You must not override anything manually in `hackage-packages.nix`. This file gets regenerated every night and will override your changes.
You should set all version overrides in `pkgs/development/haskell-modules/configuration-common.nix`.

maralorn commented on pkgs/development/misc/haskell/hasura/graphql-engine/default.nix:
```diff
@@ -2,11 +2,11 @@
 , ci-info, containers, criterion, cryptonite, data-has, deepseq
-, dependent-map, dependent-sum, directory, ekg-core, ekg-json
+, dependent-map_0_2_4_0, dependent-sum_0_4, directory, ekg-core, ekg-json
 , fast-logger, fetchgit, file-embed, filepath, generic-arbitrary
 , ghc-heap-view, graphql-parser, hashable, hspec, hspec-core
 , hspec-expectations-lifted, http-client, http-client-tls
-, http-types, immortal, insert-ordered-containers, jose, lens
+, http-types, immortal_0_2_2_1, insert-ordered-containers, jose, lens
```
It would be much more recommended to make a clean split here.
This file should just be the output generated by cabal2nix.

All manual overrides should happen in `pkgs/development/haskell-modules/configuration-common.nix`. That way the next time you update this file you don‘t need to manually insert version overrides again.

maralorn commented on pkgs/development/misc/haskell/hasura/graphql-parser/default.nix:
```diff
@@ -1,34 +1,35 @@
 { mkDerivation, aeson, attoparsec, base, bytestring, containers
 , criterion, fetchgit, filepath, hedgehog, hpack, prettyprinter
-, protolude, regex-tdfa, scientific, stdenv, template-haskell, text
+, protolude_0_3_0, regex-tdfa, scientific, stdenv, template-haskell, text
```
Same as with the other files.

maralorn commented on pkgs/servers/hasura/default.nix:
```diff
@@ -4,13 +4,12 @@ with haskell.lib;
+  version = "1.3.1-beta";
 
-  pkgs = haskell.packages.ghc865.override {
+  pkgs = haskell.packages.ghc8101.override {
```
Out of curiosity. Is it important that you build this with 8.10 and not with 8.8 which is the current default in nixpkgs?

That's no deal breaker, but It would have the advantage of much more dependency reuse.

maralorn replied:
GHC 8.8 will stay the default compiler for the 20.09 release, btw.

maralorn reviewed (changes requested):
Hey, I guess peti is a bit busy, so I hope it's okay when I jump in here.

I have quite a few improvement suggestions, but I think it's worth it. Feel free to ask if you need any more guidance. I`m glad to help.

If we have everything cleared up a merge won‘t be a problem I think.

maralorn commented:
Also all Haskell related PR's should go against the haskell-updates branch. I took the liberty to fix that.

ingenieroariel commented:
This is pretty awesome feedback, I'll submit a new version in a few days.

Thanks @malarorn

On Fri, 14 Aug 2020 at 5:58 PM maralorn <notifications@github.com> wrote:

> *@maralorn* commented on this pull request.
> ------------------------------
>
> In pkgs/servers/hasura/default.nix
> <https://github.com/NixOS/nixpkgs/pull/95369#discussion_r470897566>:
>
> >
> -  pkgs = haskell.packages.ghc865.override {
> +  pkgs = haskell.packages.ghc8101.override {
>
> GHC 8.8 will stay the default compiler for the 20.09 release, btw.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/95369#discussion_r470897566>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAANNVYTXFNKMFZOPPCYCRDSAW6QRANCNFSM4P62W7GQ>
> .
>

cdepillabout commented:
I don't think I saw this mentioned above, but I'd also recommend you urge the author of `hasura` to release to Hackage.  It is often easier for distros (and Nixpkgs in specific) to package Haskell packages when they come from Hackage.

maralorn reviewed (changes requested):
Nice, we are getting there. Just a few more improvements.

maralorn reviewed (changes requested):
I have just a few minor remarks left. But the main point is, that it would be nice if you could regenerate the hackage-packages.nix so we can test this thoroughly.

If you have trouble regenerating the hackage-packages.nix I can try to do it.

maralorn commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -1445,6 +1445,39 @@ self: super: {
+    optics-extra = self.optics-extra_0_3;
+  };
+
+  list-t = dontCheck self.list-t;
+
+  superbuffer = dontCheck (doJailbreak self.superbuffer);
+
+  Spock-core = dontCheck self.Spock-core;
+ 
+  stm-containers = dontCheck self.stm-containers;
+
+  stm-hamt = doJailbreak self.stm-hamt;
+
+  monoidal-containers = doJailbreak self.monoidal-containers;
```
Oh, wow. I use that one, but I hadn‘t dreamed of getting it into nixpkgs. Good luck with that.

ingenieroariel replied:
offtopic: I am thinking on splitting the problem in two: Haskell only (doable) and Nix only (nearly impossible).

ingenieroariel commented:
I added comments for the disabled tests (turns out it is all related to HTF being broken and those packages using it for their tests), squashed into one commit, regenerated everything using cabal2nix and committed the updated files after verifying it built locally.

ingenieroariel commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -1445,6 +1445,39 @@ self: super: {
+
+  stm-hamt = doJailbreak self.stm-hamt;
+
+  monoidal-containers = doJailbreak self.monoidal-containers;
```
It is needed because the problem is with base on GHC 8.10

ingenieroariel commented:
I need some advice managing the HFT issue for GHC 8.10 that I guess is not present in GHC 8.8

ingenieroariel commented:
@ofborg eval

ingenieroariel commented:
Removed a bunch of people that had been added to the list of reviewers by mistake after pushing the wrong branch (master instead of haskell-updates).

ingenieroariel commented:
I have been testing the hasura-graphql-engine produced by the latest version of this PR for a while and it seem to work fine. @maralorn or @offlinehacker if you could take it for a spin and verify it works for you it would be really appreciated.

@peti I hope this version is more in line with your original idea, thanks for the patience.

offlinehacker commented on pkgs/development/haskell-modules/non-hackage-packages.nix:
```diff
@@ -23,4 +23,9 @@ self: super: {
   haskell-language-server = self.callPackage ../tools/haskell/haskell-language-server { };
   hls-ghcide = self.callPackage ../tools/haskell/haskell-language-server/hls-ghcide.nix { };
+
+  ci-info = self.callPackage ../misc/haskell/hasura/ci-info {};
```
Maybe give comments how these files were generated, so it's easier to update next time?

ingenieroariel replied:
Will leave your original comments, good idea - thanks!

offlinehacker reviewed (approved):
I just gave a few suggestions, but everything else looks good to me. This solution is much better than my initial solution and I also learned much from it. It's up to Haskell maintainers now, to say if they are ok with this.

peti commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -1414,6 +1414,42 @@ self: super: {
+     dependent-map = self.dependent-map_0_2_4_0;
+     dependent-sum = self.dependent-sum_0_4;
+     witherable = self.witherable_0_3_2;
+  }).overrideDerivation(d: rec {
```
Please don't use `overrideDerivation`. That function is a hack and should be avoided as much as possible. The proper solution here is `overrideCabal`.

maralorn replied:
This should be easy to fix.

ingenieroariel replied:
Done

peti commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -1414,6 +1414,42 @@ self: super: {
+  };
+
+  # Hasura 1.3.1
+  graphql-engine = (super.graphql-engine.override {
```
Why don't you configure these overrides in `pkgs/development/haskell-modules/non-hackage-packages.nix` where you define the package?

maralorn replied:
I am sorry, that I didn‘t realize this before. But I agree, that it would be better to do those overrides directly at the point where you you include the modules. Sorry, that I didn‘t know.

ingenieroariel replied:
There is a comment at the top of that file that reads:

```
# This file should only contain packages that are not in ./hackage-packages.nix.
# Attributes in this set should be nothing more than a callPackage call.
# Overrides to these packages should go to either configuration-nix.nix,
# configuration-common.nix or to one of the compiler specific configuration
# files.
```
I assume this is in case it gets to hackage things continue working. Doing the requested changes goes against that comment, should I still do them?

maralorn replied:
Huh, I would leave that one for @peti to decide. But I think the comment makes sense. Imagine hasura being released to hackage without any code changes. Then we would still need the overrides in configuration-common.nix.

ingenieroariel replied:
/cc @basvandijk

ingenieroariel replied:
Copying the author of the comment to clarify - both options seem sensible to me.

peti commented on pkgs/development/haskell-modules/configuration-hackage2nix.yaml:
```diff
@@ -2615,7 +2615,11 @@ extra-packages:
   - yesod-persistent < 1.5              # pre-lts-11.x versions neeed by git-annex 6.20180227
   - yesod-static ^>= 1.5                # pre-lts-11.x versions neeed by git-annex 6.20180227
   - yesod-test ^>= 1.5                  # pre-lts-11.x versions neeed by git-annex 6.20180227
-
```
Please don't delete that empty line.

ingenieroariel replied:
Fixed.

peti commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -1414,6 +1414,42 @@ self: super: {
+     witherable = self.witherable_0_3_2;
+  }).overrideDerivation(d: rec {
+     # version in cabal file is invalid
+     version = "1.3.1-beta1";
```
What makes you say that the version number would be invalid? If I'm not mistaken, then it's `1.0.0`, which looks like a perfectly valid version number.

offlinehacker replied:
Their release versions are different, compared to what is specified in cabal file: https://github.com/hasura/graphql-engine/releases They are also using these versions as `VERSION` env variable during build, so they don't use version in cabal file at all. This gives only one conclusion, that they don't update version in cabal file, and yes consequently we need to patch version, regardless whether is wrong or not.

peti replied:
OK, I see. Upstream is really jumping through all kinds of hoops here to make sure that people who distribute their software have a maximum hard time. :-(

maralorn commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -1236,10 +1235,11 @@ self: super: {
   # Bump these version bound
-  monoidal-containers = appendPatch super.monoidal-containers (pkgs.fetchpatch {
+  # For GHC 8.10 base constraint also needs relaxing.
+  monoidal-containers = doJailbreak (appendPatch super.monoidal-containers (pkgs.fetchpatch {
```
The compiler version specific jailbreaks should go to configuration-ghc-8.10.x.nix.

ingenieroariel replied:
I tried but I still got the error - improved the comment to mention that.

maralorn replied:
Strange, I'm gonna investigate that.

maralorn replied:
When I enter `monoidal-containers = doJailbreak super.monoidal-containers;` into `configuration-ghc-8.10.x.nix`
`haskell.packages.ghc8101.monoidal-containers` builds.

maralorn replied:
So maybe you can try that again?

maralorn commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -1414,6 +1414,42 @@ self: super: {
+
+  # Tests depend on HTF and it is broken, 2020-08-23
+  list-t = dontCheck super.list-t;
+  superbuffer = dontCheck (doJailbreak super.superbuffer);
```
Do you have a comment as to why wee need a jailbreak here?

ingenieroariel replied:
All of them have the same root cause. I just updated it to have the same comment each time so it is not lost on a refactor.

maralorn replied:
Well, sorry, for nitpicking. You don‘t need to repeat the comment, I think. `dontCheck` is not a huge sin. I am more worried about the `doJailbreak` and I don‘t understand how your comment addresses that. Can you make that a bit more explicit?

ingenieroariel replied:
I removed the doJailbreak and it compiled, taking it out of the PR

maralorn reviewed (changes requested):
So many comments by so many people. ^^ All of them look sensible. This will be one great PR, when it's ready. ;-)

ingenieroariel commented on pkgs/top-level/all-packages.nix:
```diff
@@ -16187,17 +16187,9 @@ in
-    hasura-cli
-    hasura-graphql-engine;
-   */
+  hasura-graphql-engine = haskell.lib.justStaticExecutables haskell.packages.ghc8101.graphql-engine;
```
A lot of other files in `all-packages.nix` use that so it is cargo culting for now. What's the alternative?

maralorn replied:
The alternative would be just not calling `justStaticExecutables` on the package. But I don‘t really care either way. If people wan‘t it as a lib, they won‘t use this top-level name, so I don‘t see the harm.

offlinehacker replied:
ignore that, it's good we only have executable and if that's a pattern used for haskell

maralorn commented:
@ingenieroariel contacted me via irc, regarding merging this PR before branch-off. I really think we can and should do this.

Provided #96560 will get merged before branch-off. I am not sure about that. If not I think once we have this merged, backporting shortly after branch-off shouldn‘t be a problem, seeing that this is a fix, that has not really a danger of breaking anything (if it evaluates).

I think the main concerns at this point are about code maintainability. I will do one last review and run some tests, but I think a merge until friday should be completely doable.

maralorn commented:
@GrahamcOfBorg build haskellPackages.HTF
@GrahamcOfBorg build haskellPackages.monoidal-containers
@GrahamcOfBorg build haskellPackages.insert-ordered-containers
@GrahamcOfBorg build haskellPackages.list-t
@GrahamcOfBorg build haskellPackages.superbuffer
@GrahamcOfBorg build haskellPackages.stm-containers
@GrahamcOfBorg build haskellPackages.Spock-core
@GrahamcOfBorg build haskellPackages.graphql-engine
@GrahamcOfBorg build haskellPackages.graphql-parser
@GrahamcOfBorg build haskellPackages.ci-info
@GrahamcOfBorg build haskellPackages.pg-client
@GrahamcOfBorg build hasura-graphql-engine
@GrahamcOfBorg build hasura-cli

maralorn commented on pkgs/servers/hasura/cli.nix:
```diff
@@ -9,7 +9,7 @@ buildGoModule rec {
 
```
Any suggestions where this file would be better? I have no opinion about it.

But I think this PR changes already a lot as it is, so I would merge even without that changed.

offlinehacker replied:
good with me, does not change any functionality and we don't have any guidelines about that

offlinehacker replied:
good with me, does not change any functionality and we don't have any guidelines about that

maralorn reviewed (changes requested):
If you address my comments (which is nearly no work, I think) I will merge before Friday, if no one objects.

maralorn commented:
We have three build errors there, that look suspicious. 
HTF, graphql-engine and insert-ordered-containers. Can you explain them to me @ingenieroariel ?

ingenieroariel commented:
Let me check if what I was building locally passes on this branch:

@GrahamcOfBorg build haskell.packages.ghc810.graphql-engine
@GrahamcOfBorg build haskell.packages.ghc810.insert-ordered-containers
@GrahamcOfBorg build haskell.packages.ghc810.HTF

ingenieroariel commented:
Ok, this is really interesting, it seems my work only applies to the executable, but the libraries I touched do not build on their own neither in ghc88 or ghc810 (for different reasons):

`nix-build -A haskell.packages.ghc8101graphql-engine`
gives
```
[169 of 187] Compiling Hasura.GraphQL.Execute.LiveQuery.Poll ( src-lib/Hasura/GraphQL/Execute/LiveQuery/Poll.hs, dist/build/Hasura/GraphQL/Execute/LiveQuery/Poll.p_o )

src-lib/Hasura/GraphQL/Execute/LiveQuery/Poll.hs:57:1: error:
    Could not find module ‘GHC.AssertNF’
    Perhaps you haven't installed the profiling libraries for package ‘ghc-heap-view-0.6.2’?
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
   |
57 | import           GHC.AssertNF
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
but
`nix-build -A hasura-graphql-engine`
gives
```
/nix/store/wrc35jvllm3d4lpc6djzrvwif0q4lvy0-graphql-engine-1.3.1-beta1
```

ingenieroariel commented:
The executable package was successful:

https://github.com/NixOS/nixpkgs/pull/95369/checks?check_run_id=1057653764

But the libraries were not successful on the default compiler (ghc88x) and the target one (ghc810x).

jonringer reviewed (commented):
causes regressions in cachix and hercules-ci-agent

```
https://github.com/NixOS/nixpkgs/pull/95369
2 packages failed to build:
cachix hercules-ci-agent

3 packages built:
hasura-cli hasura-graphql-engine neuron-notes
```
cc @domenkozar 

failure:
```
builder for '/nix/store/x74iz00q1rzqbnf3if0wjvs8in9rf1w8-optics-extra-0.3.drv' failed with exit code 1; last 10 log lines:
    die', called at libraries/Cabal/Cabal/Distribution/Simple/Configure.hs:1022:20 in Cabal-3.0.1.0:Distribution.Simple.Configure
    configureFinalizedPackage, called at libraries/Cabal/Cabal/Distribution/Simple/Configure.hs:475:12 in Cabal-3.0.1.0:Distribution.Simple.Configure
    configure, called at libraries/Cabal/Cabal/Distribution/Simple.hs:625:20 in Cabal-3.0.1.0:Distribution.Simple
    confHook, called at libraries/Cabal/Cabal/Distribution/Simple/UserHooks.hs:65:5 in Cabal-3.0.1.0:Distribution.Simple.UserHooks
    configureAction, called at libraries/Cabal/Cabal/Distribution/Simple.hs:180:19 in Cabal-3.0.1.0:Distribution.Simple
    defaultMainHelper, called at libraries/Cabal/Cabal/Distribution/Simple.hs:116:27 in Cabal-3.0.1.0:Distribution.Simple
    defaultMain, called at Setup.hs:4:8 in main:Main
  Setup: Encountered missing or private dependencies:
  optics-core >=0.3 && <0.3.1
cannot build derivation '/nix/store/hxh2fn8vb3i2lr9rskv7b0pw8w279ys0-insert-ordered-containers-0.2.3.1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/1hvhik5ng6cxpiimhycw4fl3cx26vryk-swagger2-2.5.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/a7c6v7fls9cb0yq1sa2p7xl6arqxcv57-servant-swagger-1.1.7.1.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/1aij6rjwx1m1kmqkmjl8d1gpxp6k7678-servant-swagger-ui-core-0.3.3.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/f23x9461awib975ji6fbya01g40fjbk5-servant-auth-swagger-0.2.10.0.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/h19a18vgqj00jyz7pdwc098prw73sllz-cachix-api-0.4.0.drv': 4 dependencies couldn't be built
cannot build derivation '/nix/store/ixbv571inhlc9bba8sxqcbw2agbxrg14-hercules-ci-api-core-0.1.1.0.drv': 4 dependencies couldn't be built
cannot build derivation '/nix/store/8adg3a8injii6rc5mhx962zs64gc8sxk-cachix-0.3.8.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/qd7d4vrglx0v0hmg1s5kxl2msxlmwjqj-cachix-0.3.8.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/i96w3d8nids2lhz0dzyiig477xlr2lgj-hercules-ci-api-agent-0.2.2.0.drv': 5 dependencies couldn't be built
cannot build derivation '/nix/store/ricjji66azz0881q3rj9rglxmk209afa-hercules-ci-agent-0.7.4.drv': 4 dependencies couldn't be built
```

maralorn commented:
Okay, this sucks, but all overrides that break packages which normally build on 8.8 need to be moved to the 8.10 file. This concerns at least `insert-ordered-containers`, which as a dep of cachix should really not get broken.

maralorn commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -297,7 +297,6 @@ self: super: {
   hsbencher = dontCheck super.hsbencher;
   hsexif = dontCheck super.hsexif;
   hspec-server = dontCheck super.hspec-server;
-  HTF = dontCheck super.HTF;
```
Wait, why do you remove this? This introduces a build error for this package. Just not changing this line could fix the HTF package. Or is there some reason I don‘t get? Maybe then you can even remove some of the `dontCheck`s below for which broken HTF is actually the reason? Or is HTF broken either way already before this PR? Anyway just let me now, what your rationale here was.

ingenieroariel replied:
I think this was a mistake, I see no rationale for removing it. Sending in a new version of this PR.

ingenieroariel commented:
Things I tested locally:
```
nix-build -A haskell.packages.ghc884.graphql-engine:
Fails with -> Module ‘System.IO.Error’ does not export ‘isResourceVanishedError’.

I tried making base = self.base-compat; but it did not work: https://hackage.haskell.org/package/base-compat-0.11.1/docs/System-IO-Error-Compat.html
```

```
nix-build -A haskellPackages.cachix
-> /nix/store/a1mjx07gni83n011b2max5hfq3qbaxff-cachix-0.3.8
```

```
nix-build -A haskellPackages.hercules-ci-agent
/nix/store/haqfir8haqwl6kx6423ib2grjf8cmjhz-hercules-ci-agent-0.7.4
```
```
nix-build -A haskell.packages.ghc8101.graphql-engine
/nix/store/jdvdaayc73fimvqy8d0545z2kmxirc9s-graphql-engine-1.3.1-beta1
```

```
nix-build -A haskellPackages.HFT
gives -> <no location info>: error:
    `cc' failed in phase `C Compiler'. (Exit code: 1)
```

ingenieroariel commented:
@GrahamcOfBorg build haskellPackages.HTF
@GrahamcOfBorg build haskellPackages.monoidal-containers
@GrahamcOfBorg build haskellPackages.insert-ordered-containers
@GrahamcOfBorg build haskellPackages.list-t
@GrahamcOfBorg build haskellPackages.superbuffer
@GrahamcOfBorg build haskellPackages.stm-containers
@GrahamcOfBorg build haskellPackages.Spock-core
@GrahamcOfBorg build haskellPackages.graphql-engine
@GrahamcOfBorg build haskellPackages.graphql-parser
@GrahamcOfBorg build haskellPackages.ci-info
@GrahamcOfBorg build haskellPackages.pg-client
@GrahamcOfBorg build hasura-graphql-engine
@GrahamcOfBorg build hasura-cli

@GrahamcOfBorg build cachix
@GrahamcOfBorg build hercules-ci-agent

@GrahamcOfBorg build haskell.packages.ghc8101.graphql-engine
@GrahamcOfBorg build haskell.packages.ghc884.graphql-engine

maralorn commented:
Okay, monoidal-containers is as of last night broken on haskell-updates, I fixed that.

@GrahamcOfBorg build haskellPackages.monoidal-containers
@GrahamcOfBorg build hasura-graphql-engine
@GrahamcOfBorg build haskellPackages.graphql-parser

ingenieroariel commented:
@GrahamcOfBorg build haskellPackages.monoidal-containers
@GrahamcOfBorg build hasura-graphql-engine
@GrahamcOfBorg build haskellPackages.graphql-parser

maralorn commented on pkgs/development/haskell-modules/configuration-ghc-8.10.x.nix:
```diff
@@ -76,6 +76,15 @@ self: super: {
+      optics-core = self.optics-core_0_3_0_1;
+    };
+  };
+
```
```suggestion

  # Jailbreaking because monoidal-containers hasn‘t bumped it's base dependency for 8.10.
```

maralorn commented:
Let's just check everything once again. Just to be sure. I guess most of it will be cached anyways.

@GrahamcOfBorg build haskellPackages.monoidal-containers
@GrahamcOfBorg build haskellPackages.insert-ordered-containers
@GrahamcOfBorg build haskellPackages.list-t
@GrahamcOfBorg build haskellPackages.superbuffer
@GrahamcOfBorg build haskellPackages.stm-containers
@GrahamcOfBorg build haskellPackages.Spock-core
@GrahamcOfBorg build haskellPackages.graphql-engine
@GrahamcOfBorg build haskellPackages.graphql-parser
@GrahamcOfBorg build haskellPackages.ci-info
@GrahamcOfBorg build haskellPackages.pg-client
@GrahamcOfBorg build hasura-graphql-engine
@GrahamcOfBorg build hasura-cli

@GrahamcOfBorg build cachix
@GrahamcOfBorg build hercules-ci-agent

@GrahamcOfBorg build haskell.packages.ghc8101.graphql-engine
@GrahamcOfBorg build haskell.packages.ghc884.graphql-engine

maralorn commented:
Result of `nixpkgs-review pr 95369` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
<br>- hasura-cli
<br>- hasura-graphql-engine
</details>

ingenieroariel commented:
@GrahamcOfBorg build haskellPackages.monoidal-containers
@GrahamcOfBorg build haskellPackages.insert-ordered-containers
@GrahamcOfBorg build haskellPackages.list-t
@GrahamcOfBorg build haskellPackages.superbuffer
@GrahamcOfBorg build haskellPackages.stm-containers
@GrahamcOfBorg build haskellPackages.Spock-core
@GrahamcOfBorg build haskellPackages.graphql-engine
@GrahamcOfBorg build haskellPackages.graphql-parser
@GrahamcOfBorg build haskellPackages.ci-info
@GrahamcOfBorg build haskellPackages.pg-client
@GrahamcOfBorg build hasura-graphql-engine
@GrahamcOfBorg build hasura-cli

@GrahamcOfBorg build cachix
@GrahamcOfBorg build hercules-ci-agent

@GrahamcOfBorg build haskell.packages.ghc8101.graphql-engine
@GrahamcOfBorg build haskell.packages.ghc884.graphql-engine

maralorn commented:
Thanks, for sticking with it 'til the end! Great work!

datakurre commented:
@ingenieroariel @maralorn Something happened to the closure size of hasura-graphql-engine in this merge. A xz packaged closure at piens/nixpkgs:2b0b078a64162ab193f2aeb1c88bd2e92f42e90c was about 20MB. Now, at nixos/nixpkgs:nixos-20.09 (d447429cc2407f7e76b09299439991d777914f4e) it is above 400MB.

maralorn commented:
Oh, well. That is a finicky thing. I guess somewhere something introduced a runtime dependency on the Haskell ecosystem.

I honestly don‘t know what the best practices are to prevent such a thing.

If we know what the issue is we can probably fix something with references-to or simpler, but somebody will have to compare those two closures and it's dependency trees to find out where the issue lies.

maralorn commented:
On second though:

This PR dropped the `justStaticExecutables` for hasura-graphql-engine. Probably it's just that? I didn‘t think about closure sizes at that time. Lesson learned.

If some could try that out, that would be great. Otherwise open an issue so we don‘t forget about it.
